src/swap.rs

```diff
@@ -0,0 +1,354 @@
+    fn spend(&self, tokens: &[Proof]) -> AnyResult<()>;
+    fn get_state(&self, ys: &[Proof]) -> AnyResult<Vec<ProofState>>;
+}
+
+pub struct Service<KeysRepo, ProofRepo> {
+    keys: KeysRepo,
+    proofs: ProofRepo,
+}
+
+impl<KeysRepo, ProofRepo> Service<KeysRepo, ProofRepo>
+where
+    KeysRepo: KeysRepository,
+    ProofRepo: ProofRepository,
+{
+    fn verify_proofs_are_unspent(&self, proofs: &[Proof]) -> Result<bool> {
+        let result = self
+            .proofs
+            .get_state(proofs)
+            .map_err(Error::ProofRepository)?
+            .into_iter()
+            .all(|state| state == ProofState::Unspent);
+        Ok(result)
+    }
+
+    fn verify_proofs_signatures(&self, proofs: &[Proof]) -> Result<bool> {
+        for proof in proofs {
+            let id = proof.keyset_id;
+            let keyset = self
+                .keys
+                .load(id.into())
+                .map_err(Error::KeysetRepository)?
+                .ok_or_else(|| Error::UnknownKeyset(id.into()))?;
+            let key = keyset
+                .keys
+                .get(&proof.amount)
+                .ok_or_else(|| Error::UnknownAmountForKeyset(id.into(), proof.amount))?;
+            let ok = cdk::dhke::verify_message(&key.secret_key, proof.c, proof.secret.as_bytes());
+            if ok.is_err() {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn swap(
+        &self,
+        inputs: &[Proof],
+        outputs: &[BlindedMessage],
+    ) -> Result<Vec<BlindSignature>> {
+        // first step: zero-cost verifications
+        let no_zero_amount = outputs.iter().all(|output| output.amount != Amount::ZERO);
+        if !no_zero_amount {
+            return Err(Error::ZeroAmount);
+        }
+        let total_input: Amount = inputs
+            .iter()
+            .fold(Amount::ZERO, |total, proof| total + proof.amount);
+        let total_output: Amount = outputs
+            .iter()
+            .fold(Amount::ZERO, |total, output| total + output.amount);
+        if total_input != total_output {
+            return Err(Error::UnmatchingAmount(total_input, total_output));
+        }
+        // second step: costly verifications
+        let proofs_are_unspent = self.verify_proofs_are_unspent(inputs)?;
+        if !proofs_are_unspent {
+            return Err(Error::ProofsAlreadySpent);
+        }
+        let proofs_signatures_are_ok = self.verify_proofs_signatures(inputs)?;
+        if !proofs_signatures_are_ok {
+            return Err(Error::UnknownProofs);
+        }
+
+        let mut ids: Vec<KeysetID> = Vec::new();
+        for i in inputs {
+            let o = self
+                .keys
+                .replacing_keysetid(i.keyset_id.into())
+                .map_err(Error::KeysetRepository)?
+                .ok_or(Error::UnknownKeyset(i.keyset_id.into()))?;
+            ids.push(o);
+        }
+        let first = ids.first().expect("first is None");
```

Comment:
Should there maybe also be a check that `inputs` can't be empty? :thinking: (if it's empty, ids is also empty I think and then this panics)